### ISSUE TITLE
`Lint/UselessAssignment` must be considered unsafe

### ISSUE THREAD

**Issue by shyouhei** [rubocop/rubocop#14445]
## Expected behavior

I don't mind if this code has offenses, but this auto-correction is WRONG; must not happen.

It is super-duper difficult -- if not impossible -- to statically prove if a variable assignment is useless or not.  Rubocop does not have such feature.  That is a reasonable, practical design decision that I support.  You should just stop pretending that you do so.

## Actual behavior

Autocorrection breaks semantics of the code.

## Steps to reproduce the problem

```console
$ cat tmp.rb
def foo
  backtrack = []
  result = []
  node = self

  until node.nil?
    children = node.children
    result << children.first

    unless children.second
      node = backtrack.pop
      next
    end

    backtrack.push(children[2..]).flatten! if children.third
    node = children.second
  end

  result
end
$
$
$ bin/rubocop --safe-auto-correct tmp.rb
--safe-auto-correct is deprecated; use --autocorrect instead.
Inspecting 1 file
W

Offenses:

tmp.rb:11:7: W: [Corrected] Lint/UselessAssignment: Useless assignment to variable - node.
      node = backtrack.pop
      ^^^^

1 file inspected, 1 offense detected, 1 offense corrected
```

## RuboCop version

```
$ bin/rubocop -V
1.79.0 (using Parser 3.3.9.0, rubocop-ast 1.46.0, analyzing as Ruby 3.2, running on ruby 3.2.4) [x86_64-linux]
  - rubocop-faker 1.1.0
  - rubocop-performance 1.21.1
  - rubocop-rails 2.25.1
  - rubocop-rspec 3.0.4
```

**Comment by Earlopain** [comment]
Can you try upgrading to 1.79.2? I believe this has already been resolved since I can't reproduce it anymore.

**Comment by shyouhei** [comment]
> Can you try upgrading to 1.79.2? I believe this has already been resolved since I can't reproduce it anymore.

Confirmed that the script above does not render offenses any longer under 1.79.2.  But I can think of other codes that still do:

```ruby
def foo
  node = self
  for node in node.children&.pop do
    result = node if node.has_children?
  end
  result
end
```

**Comment by Earlopain** [comment]
Thanks for the repro, I openend https://github.com/rubocop/rubocop/pull/14458 to fix this.
